File: pathfinding.js
var map = [
	[
		'user',
		'empty',
		'empty',
		'empty',
		'song'
	],
	[
		'wall',
		'wall',
		'wall',
		'wall',
		'empty'
	],
	[
		'empty',
		'empty',
		'empty',
		'empty',
		'empty'
	],
	[
		'empty',
		'wall',
		'wall',
		'wall',
		'wall'
	],
	[
		'empty',
		'empty',
		'empty',
		'empty',
		'monkey'
	]
];

function floodfill(map, x, y, value) {

	var res = map.map(function(arr) {
		return new Array(arr.length);
	});

	var queue = [{x: x, y: y, value: 0}];
	while(queue.length > 0) {
		var n = queue.pop();

		if(map[n.y] === undefined ||
				map[n.y][n.x] === undefined) {
			continue;
		}
		else if(res[n.y] === undefined ||
			res[n.y][n.x] === undefined) {

			if(map[n.y][n.x] !== 'empty' &&
				map[n.y][n.x] !== 'monkey') {
				continue;
			}

			res[n.y][n.x] = {
				type: map[n.y][n.x],
				value: n.value,
				dir: n.dir,
				parent: n.parent
			};
		
			queue.push({x: n.x, y: n.y-1, value: n.value+1, parent: {x: n.x, y: n.y}, dir: 'up'});
			queue.push({x: n.x+1, y: n.y, value: n.value+1, parent: {x: n.x, y: n.y}, dir: 'right'});
			queue.push({x: n.x, y: n.y+1, value: n.value+1, parent: {x: n.x, y: n.y}, dir: 'down'});
			queue.push({x: n.x-1, y: n.y, value: n.value+1, parent: {x: n.x, y: n.y}, dir: 'left'});
		}
	}
	return res;
}

function backtrace(map, x, y) {
	if(map[y][x] === undefined) {
		return [];
	}
	var p = map[y][x].parent;

	if(p === undefined) {
		return [];
	}

	var c = map[p.y][p.x];
	var path = [];
	var last = {x: x, y: y};

	while(c.parent !== undefined) {
		p = c.parent;
		c = map[p.y][p.x];

		path.push(c.dir);
		last = p;
	}
	path.reverse();
	return path;
}

function prettymap(map) {
	var rows = "";
	for(var y = 0; y < map.length; y++) {
		var line = "";
		for(var x = 0; x < map[y].length; x++) {
			switch(map[y][x]) {
				case 'empty': line += '. '; break;
				case 'wall': line += '# '; break;
				case 'song': line += 'S '; break;
				case 'playlist': line += 'P '; break;
				case 'album': line += 'A '; break;
				case 'banana': line += 'B '; break;
				case 'monkey': line += 'M '; break;
				case 'user': line += 'U '; break;
			}
		}
		rows += line + '\n';
	}
	return rows;
}


console.log(prettymap(map));
var res = floodfill(map, 4, 4, 0);
console.log(res);
console.log(backtrace(res, 4, 1));


exports.floodfill = floodfill;
exports.backtrace = backtrace;
exports.prettymap = prettymap;

File: src/mmc2/pathfinding.clj
(ns mmc2.pathfinding
  (:require [clojure.set]))

(def directions
  {[ 1  0] "down" 
   [-1  0] "up"   
   [ 0  1] "right"
   [ 0 -1] "left"})

(def points
  {"song" 1
   "album" 2
   "playlist" 4
   "banana" 5
   "trap" 0
   "user" -1
   "monkey" 0})


(defn translate-coords
  [[x y] [dx dy]]
  [(+ x dx) (+ y dy)])

(defn neighbour-coords
  [coord]
  (map translate-coords (repeat coord) (keys directions)))

(defn walkable
  [layout coord]
  (let [item (get-in layout coord "wall")]
    (not= "wall" item)))
    ;(or (= "empty" item) (= "monkey" item))))

(defn neighbours
  [layout coord]
  (if (walkable layout coord)
    (let [walkable-coords (filter #(walkable layout %) (neighbour-coords coord))]
      (zipmap walkable-coords (repeat coord)))
    ()))

(defn backtrace-path
  [visited coord]
  (loop [path [coord]
         c coord]
    (let [parent (visited c)]
      (if (nil? parent)
        path
        (recur (conj path parent) parent)))))

(defn path-to-dirs
  [path]
  (loop [[f s & xs] path
         dirs []]
    (if (seq s)
      (let [dirs (conj dirs (directions [(- (f 0) (s 0)) (- (f 1) (s 1))]))]
        (recur (cons s xs) dirs))
      (rseq dirs))))


(defn flood-fill
  [layout coord]
  (loop [visited {}
         frontier {coord nil}
         paths []]
    (if (seq frontier)
      (let [current (first frontier)
            item (get-in layout (key current))
            frontier (dissoc frontier (key current))
            visited (merge visited current)
            candidates (neighbours layout (key current))
            frontier (if (or (= "empty" item) (= "monkey" item))
                       (merge frontier (apply dissoc candidates (keys visited)))
                       frontier)
            back (backtrace-path visited (key current))
            paths (if (and (not= "empty" item) (not= "monkey" item))
                    (conj paths {:coord(key current)
                                 :path (path-to-dirs back)
                                 :points (points item)})
                    paths)]
        (recur visited frontier paths))
      paths)))

